src/stcal/outlier_detection/median.py

```diff
@@ -0,0 +1,355 @@
+"""Compute median of large datasets in memory- and runtime-efficient ways."""
+
+from __future__ import annotations
+
+import logging
+import tempfile
+import warnings
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+_ONE_MB = 1 << 20
+
+
+def nanmedian3D(cube: np.ndarray, overwrite_input: bool = True) -> np.ndarray:
+    """Compute the median of a cube ignoring warnings and with
+    memory efficiency optimizations. np.nanmedian always uses at least 64-bit
+    precision internally, and this is too memory-intensive. Instead, loop over
+    the median calculation to avoid the memory usage of the internal upcasting
+    and temporary array allocation. The additional runtime of this loop is
+    indistinguishable from zero, but this loop cuts overall step memory usage
+    roughly in half for at least one test association.
+    """
+    with warnings.catch_warnings():
+        warnings.filterwarnings(action="ignore",
+                                message="All-NaN slice encountered",
+                                category=RuntimeWarning)
+        output_arr = np.empty(cube.shape[1:], dtype=np.float32)
+        for i in range(output_arr.shape[0]):
+            # this for loop looks silly, but see docstring above
+            np.nanmedian(cube[:, i, :],
+                         axis=0,
+                         overwrite_input=overwrite_input,
+                         out=output_arr[i, :])
+        return output_arr
+
+
+class MedianComputer:
+    """
+    Top-level class to treat median computation uniformly, whether in
+    memory or on disk.
+    """
+
+    def __init__(self: MedianComputer,
```

Comment:
```suggestion
    def __init__(self: MedianComputer,
```
Is annotating `self` common practice for type hinting? I imagine it would make class renaming difficult and what would it do for a subclass?

Reply:
I didn't think it was necessary either, but then mypy was complaining locally.  I'll check if this can be removed setting mypy up to take its configuration from the stcal pyproject.toml

Reply:
Perhaps `from typing import Self`? (only works for python 3.11+ maybe that's enough for our CI checks).

Reply:
I was wrong about the failures being from mypy, they were actually caused by our ruff check.  See the small change I made to the pyproject.toml: [ANN101](https://docs.astral.sh/ruff/rules/missing-type-self/) and [ANN102](https://docs.astral.sh/ruff/rules/missing-type-cls/) are specific to `self` and `cls` annotations, and both are deprecated